### PR TITLE
feat(decimal): number formatting for different locales - FE-3500

### PR DESCRIPTION
**Pull request by maspychaj-sage**
### Proposed behaviour
Added props to allow different formatting numbers (custom delimiter and separator).

### Current behaviour
Only delimiter with comma and separator with dot was used. By default number wasn't formatted differently on change locale.

### Checklist

- [x ] Commits follow our style guide
<del>- [ ] Screenshots are included in the PR if useful</del>
- [x ] All themes are supported if required
- [x ] Unit tests added or updated if required
<del>- [ ] Cypress automation tests added or updated if required</del>
- [x ] Storybook added or updated if required
<del>- [ ] Typescript `d.ts` file added or updated if required</del>
<del>- [ ] Carbon implementation and Design System documentation are congruent</del>

### CODE REVIEW

**Comment by sjohnson-sage** [comment]
Please don't move this PR out of draft until it has two Carbon approvals - this is to stop Chromatic running on code that is likely to change and wasting snapshots

**Comment by sjohnson-sage** [comment]
It looks like there's been a bad rebase with master from the commits on this PR, can you sort it out please @maspychaj-sage?

**Review by alzieba — APPROVED** [review]

**Review by PaulSumner-Sage — DISMISSED** [review]

File: src/__experimental__/components/decimal/decimal.stories.mdx
Comment (chrisbarber86) [inline src/__experimental__/components/decimal/decimal.stories.mdx:286]: Why do we need to pass in `locale` AND `delimiter`/`separator`. If we are being told the delimiter and the separator then we shouldn't need the locale (or vice versa)

File: src/__experimental__/components/decimal/decimal.component.js
Comment (chrisbarber86) [inline src/__experimental__/components/decimal/decimal.component.js:188]: Why are we overwriting the translations for the locale? That's not something this component should be doing.

if the component receives `locale` then we should use I18n.t to get the format. Or we remove local and we just use the provided props.

**Review by chrisbarber86 — CHANGES_REQUESTED** [review]

File: src/__experimental__/components/decimal/decimal.stories.mdx
Comment (maspychaj-sage) [reply src/__experimental__/components/decimal/decimal.stories.mdx:286]: I used the same approach as in code example: https://codesandbox.io/s/numberformatting-kijbl?file=/src/locale.js - we can change it to use delimiter and separator only.

File: src/__experimental__/components/decimal/decimal.component.js
Comment (maspychaj-sage) [reply src/__experimental__/components/decimal/decimal.component.js:188]: @chrisbarber86  - removing usage of I18n.locale requires also changes for toStandardDecimal - just using delimiter and separator from props in formatDecimal functions cause change of hiddenValue. Do you really think that'll be better in that case?

File: src/__experimental__/components/decimal/decimal.stories.mdx
Comment (chrisbarber86) [reply src/__experimental__/components/decimal/decimal.stories.mdx:286]: I think i would prefer that we expect delimiter and separator to be passed in. Then the consuming application can use whatever library they want for I18n.

File: src/__experimental__/components/decimal/decimal.component.js
Comment (chrisbarber86) [reply src/__experimental__/components/decimal/decimal.component.js:188]: my point was you are changing the value of I18n.locale AND I18n.translations[I18n.locale] - both of which are defined and owned by the application - this component has no right to override them. A requirement of using I18n here would be that the consuming application knows to setup:

```javascript
    {
      number: {
        format: {
          delimiter: ",",
          separator: ".",
        },
      }
    }
```

Comment (maspychaj-sage) [reply src/__experimental__/components/decimal/decimal.component.js:188]: In that case there is any better option to decide how number will be formatted? Right now it works according to browser settings and can't be changed, am I right?

**Review by jamime — CHANGES_REQUESTED** [review]
I think this has missed the intended purpose of FE-3500.

There is an associated GitHub issue that has more information https://github.com/Sage/carbon/issues/3559.

The goal is to support the `es-ES` and `pt-PT` locales that only display a delimiter once there is five or more integers.

There is more detail in the linked issue. 

Once this task is complete the decimal component should not use `i18n-js` for decimal formatting. It should use [`Intl.NumberFormat`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat).

You also need access to the delimiter and separator to allow/prevent keypresses. You can get this by calling `Intl.NumberFormat().formatToParts()`. There is another PR in flight that changes this component to allow any input, at that point you won't need to do this.
I think we should wait until that PR is merged before working on this as there is too much overlap.

Feel free to DM me on slack if you have any questions.

**Review by PaulSumner-Sage — DISMISSED** [review]

**Review by alzieba — DISMISSED** [review]

File: src/__experimental__/components/decimal/decimal.component.js
Comment (DipperTheDan) [inline src/__experimental__/components/decimal/decimal.component.js:141]: From 141 to 152 isn't required as this block of code was moved to the `toStandardDecimal` function. Here is the link to where it is now.
https://github.com/Sage/carbon/pull/3799/files#diff-41d1ef079632dd8b6d4e6270f83eb09c1dcfc8017b6ee137810ea9d3d1f41dc2R200

File: src/__experimental__/components/decimal/decimal.spec.js
Comment (DipperTheDan) [inline src/__experimental__/components/decimal/decimal.spec.js:719]: Test will need a different description as the value isn't cropped and the value shouldn't be cropped either.

File: src/__experimental__/components/decimal/decimal.spec.js
Comment (DipperTheDan) [inline src/__experimental__/components/decimal/decimal.spec.js:787]: This test has a different description but looks exactly the same as the test above?

**Review by DipperTheDan — COMMENTED** [review]
Left some comments but you should also update the `decimal.stories.js` file as that has a `locale` story. Maybe update it for the purposes of your PR or it might be easier to include a knob in the default story for the new `locale` prop.

File: src/__experimental__/components/decimal/decimal.spec.js
Comment (maspychaj-sage) [reply src/__experimental__/components/decimal/decimal.spec.js:787]: It's not the same case - defaultValue vs value - second test is for changing after defaultValue is not passed.

File: src/__experimental__/components/decimal/decimal.spec.js
Comment (DipperTheDan) [inline src/__experimental__/components/decimal/decimal.spec.js:120]: We still need this test to cover that the styled-system margin props are applied to Decimal.

**Review by DipperTheDan — COMMENTED** [review]
Just one small comment about a styled-system test you have deleted accidentally. 

It's just occurred to me in this review that you will need to add the new `locale` prop to the `decimal.d.ts` file too. Apologies I didn't pick up on that sooner! 😢

File: src/__experimental__/components/decimal/decimal.component.js
Comment (jamime) [inline src/__experimental__/components/decimal/decimal.component.js:148]:
```suggestion
    const delimiterMatcher = new RegExp(`[\\${this.getSeparator("group")} ]*`, "g");
```

lets combine these lines :)

File: src/__experimental__/components/decimal/decimal.component.js
Comment (jamime) [inline src/__experimental__/components/decimal/decimal.component.js:149]:
```suggestion
    return value.replace(delimiterMatcher, "");
```

File: src/__experimental__/components/decimal/decimal.component.js
Comment (jamime) [inline src/__experimental__/components/decimal/decimal.component.js:213]: This doesn't work for `es-ES` or `pt-PT`, these are the locales in the original issue, I think you need to change the value to `10000.1`
```suggestion
    const numberWithGroupAndDecimalSeparator = 10000.1;
```

File: src/__experimental__/components/decimal/decimal.stories.js
Comment (jamime) [inline src/__experimental__/components/decimal/decimal.stories.js:99]: Please make this configurable using a knob.

File: src/__experimental__/components/decimal/decimal.stories.js
Comment (jamime) [inline src/__experimental__/components/decimal/decimal.stories.js:139]:
```suggestion
  name: "locale",
```

File: src/utils/helpers/i18n/i18n.js
Comment (jamime) [inline src/utils/helpers/i18n/i18n.js:61]: Please restore this, this could be consumed by other projects. Looks like we don't have a unit test for this functionality. Please can you add one.

**Review by jamime — CHANGES_REQUESTED** [review]
Please can you add unit tests for the `es-es` and `pt-pt` locales. The linked issue has examples of the expected results. The current PR does not support these locales.

File: src/utils/helpers/i18n/i18n.js
Comment (jamime) [reply src/utils/helpers/i18n/i18n.js:61]: Unfortunately it is not in a `__internal__` folder so we consider it part of the public API. This helper is available as 
```js
import {formatDecimal} from "carbon-react/lib/utils/helpers/i18n";
```

In the short term we plan to remove these helpers so I think it's OK to remove it like you have as long as it's documented in the commit as a breaking change. 

I think the commit message should be something like this


```
feat(decimal): add locale support

BREAKING CHANGE: The decimal component no longer uses `i18n-js` for decimal formatting, you should use the `locale` prop instead. 

BREAKING CHANGE: `formatDecimal` no longer supports the `round` option first introduced in v73.0.0

closes FE-3500
```

**Review by jamime — CHANGES_REQUESTED** [review]
1. Using the `pt-pt` locale and entering `123 456 789` the value is not automatically formatted. In `en` we convert `123,456,789` to `123,456,789.00`

1. With en if the user enters `12,34` we will correct the delimiter `1,234`. This did not work for `pt-pt`.

Please ensure unit tests are added for these cases.

File: src/__experimental__/components/decimal/decimal.spec.js
Comment (jamime) [inline src/__experimental__/components/decimal/decimal.spec.js:623]:
```suggestion
        ])("format %s to %s", (a, b) => {
```

File: src/__experimental__/components/decimal/decimal.spec.js
Comment (jamime) [inline src/__experimental__/components/decimal/decimal.spec.js:629]:
```suggestion
        it("handles a value that has white-spaces", () => {
```

Please also add a test for 

1. the space being in wrong location.

e.g. `1 0000,00` it should correct it.

2. two spaces

e.g. `10 000` it should not format it.

File: src/__experimental__/components/decimal/decimal.spec.js
Comment (jamime) [inline src/__experimental__/components/decimal/decimal.spec.js:1031]: Why did this change?

**Review by jamime — CHANGES_REQUESTED** [review]
When using the `en` local the component should not remove any whitespace, instead it should not format the number so the consumers can display a validation message.

File: src/__experimental__/components/decimal/decimal.spec.js
Comment (jamime) [inline src/__experimental__/components/decimal/decimal.spec.js:645]: This value should not be formatted, please see previous review.

**Review by jamime — CHANGES_REQUESTED** [review]
Value with multiple delimiters should not be formatted, please compare with en and see comments on previous review.

Please remove leading space from `it` blocks.

**Review by jamime — DISMISSED** [review]

File: src/__experimental__/components/decimal/decimal.spec.js
Comment (jamime) [inline src/__experimental__/components/decimal/decimal.spec.js:623]: Please remove leading space.

File: src/__experimental__/components/decimal/decimal.spec.js
Comment (jamime) [inline src/__experimental__/components/decimal/decimal.spec.js:607]: Please remove leading space.

**Review by jamime — DISMISSED** [review]

**Review by DipperTheDan — DISMISSED** [review]

File: src/__experimental__/components/decimal/decimal.spec.js
Comment (DipperTheDan) [inline src/__experimental__/components/decimal/decimal.spec.js:615]: I think it would be great if you could add a unit test that covers `1,1,1,1` not being converted to `1.1.1.1` with the `pt-PT` locale, as per the bug @DlgSHi found. It means when we come to refactor the component to be a functional component in the near future, we have a unit test that already covers this scenario.

**Review by DipperTheDan — DISMISSED** [review]

**Review by DipperTheDan — APPROVED** [review]

**Review by DlgSHi — DISMISSED** [review]

**Review by jamime — CHANGES_REQUESTED** [review]
Lets add the same test cases that are in cypress but by testing the `rawValue` instead of `hiddenValue`.

https://github.com/Sage/carbon/blob/b9eadd94c1794e02a33ad054139fb6c30930eae1/cypress/features/regression/experimental/decimalLocale.feature#L10-L20

Please duplicate this test for each case, you can use `it.each`.

https://github.com/Sage/carbon/blob/b9eadd94c1794e02a33ad054139fb6c30930eae1/src/__experimental__/components/decimal/decimal.spec.js#L861-L874

We should also have these test cases from the original issue
https://github.com/Sage/carbon/issues/3559

| Number | Portuguese | Spanish |
| ---------------- | ---------------- | ---------------- |
| 100     | 100 | 100 |
| 1000     | 1000 | 1000 |
| 1000.23     | 1000,23 | 1000.23 |
| 10000     | 10 000 | 10.000|
| 10000.23     | 10 000,23 | 10.000,23 |
| 100000     | 100 000 | 100.000 |
| 1000000     | 1 000 000 | 1.000.000 |

Please do not rebase instead push a `test(decimal): add locale tests` commit to make reviewing easier.

**Review by DlgSHi — APPROVED** [review]

**Review by DipperTheDan — APPROVED** [review]

**Review by jamime — APPROVED** [review]

**Pull request merged**
